pick a better random port to listen on in integration tests

jboyd01 opened the pull request:
Trying to reduce flakes in CI where we randomly pick a port to listen on.  If we are unluckly its already in use and we fail the test.  

Let the OS pick a free random port to listen on.  We can set the listener early and the apiserver framework will reuse it.

MHBauer commented:
Ideally the issue is resolved by upstream letting us use the standard convention of using `0` which means "pick an available port".

n3wscott commented:
The travis build failed with 

```
    --- FAIL: TestCreateServiceInstanceWithProvisionFailure/non-url_transport_error (2.01s)
    	framework.go:83: Starting server on port: 53926
    	framework.go:115: Error in bringing up the server: failed to create listener: failed to listen on 0.0.0.0:53926: listen tcp 0.0.0.0:53926: bind: address already in use
```

There are patterns that catch this error and keep trying to pick a port until the server starts (or three fails or whatever)

Maybe it is now time to add that type of logic?

jboyd01 commented:
yep, I agree, let me give this a shot.  I debated about it originally and figured the range was wide enough, what's the chance? ;)

jboyd01 commented:
@MHBauer, @n3wscott  I've reworked, would you take another look?  When I force it to fail (ie override the initial random port and set it to 2380 where etcd is already running) I can see it initially fail and then on the 2nd attempt it always successfully listens on the new random port and the tests are successful.  I think the vulnerability we have is that if it happens to pick a random port where another controller is already listening we'll have errors as the test client thread will likely connect ok and run the test.  But that's not much of a random number generator then.  I've added plenty of logging that should make failures of this kind pretty obvious if they should happen in a test.

jberkhahn commented on test/integration/framework.go:
```diff
@@ -109,24 +120,45 @@ func withConfigGetFreshApiserverAndClient(
 		if err := server.RunServer(options, stopCh); err != nil {
-			close(serverFailed)
-			t.Logf("Error in bringing up the server: %v", err)
+			for err != nil {
```
I would add some sort of limit here just to prevent tests from hanging in CI somewhere until someone comes to look at them. 10 should be more than enough?

jboyd01 replied:
Thanks for the review @jberkhahn, I agree.  With the find from @MHBauer we get rid of this kludginess and let the OS pick the port and no more retries.

MHBauer commented on test/integration/framework.go:
```diff
@@ -109,24 +120,45 @@ func withConfigGetFreshApiserverAndClient(
+	// above hits a port already in use.
+	var controllerUrl string
+
+	if err := waitForApiserverUp(serverFailed, func() string {
```
If there is in fact a collision, what prevents this from succeeding with the wrong client before the port number is switched out?

jboyd01 commented:
Nice find @MHBauer re https://github.com/kubernetes-incubator/service-catalog/pull/1844#pullrequestreview-105189914!  Much to my surprise that works for us as well.  Much method of getting the test api server on a random port.

MHBauer commented on test/integration/framework.go:
```diff
@@ -69,45 +70,56 @@ func withConfigGetFreshApiserverAndClient(
+	if serverstorage.StorageTypeEtcd == serverConfig.storageType {
+		etcdOptions = server.NewEtcdOptions()
+		etcdOptions.StorageConfig.ServerList = serverConfig.etcdServerList
+		etcdOptions.EtcdOptions.StorageConfig.Prefix = fmt.Sprintf("%s-%08X", server.DefaultEtcdPathPrefix, rand.Int31())
```
curiosity question, why is it called int31 to get in int32? Is it immediately obvious to everyone else?

jboyd01 replied:
Not without reviewing the Int31() function comment:  "Int31 returns a non-negative pseudo-random 31-bit integer as an int32."

MHBauer commented on test/integration/framework.go:
```diff
@@ -127,6 +139,7 @@ func withConfigGetFreshApiserverAndClient(
 	if nil != err {
 		t.Fatal("can't make the client from the config", err)
 	}
+	t.Logf("Test client will use API Server URL of %v", secureAddr)
```
I feel like this gets printed in waitForAPIserverUp, but I suppose it's fine to print it as the final exit message.
